File: packages/mobile/lib/models/schedule_session.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:cloud_firestore/cloud_firestore.dart';

part 'schedule_session.freezed.dart';
part 'schedule_session.g.dart';

@freezed
class Location with _$Location {
  const factory Location({
    required double latitude,
    required double longitude,
    required double accuracy,
  }) = _Location;

  factory Location.fromJson(Map<String, dynamic> json) => _$LocationFromJson(json);
}

@freezed
class DeviceInfo with _$DeviceInfo {
  const factory DeviceInfo({
    required String deviceId,
    required String deviceModel,
    required String osVersion,
    required String appVersion,
    required double batteryLevel,
  }) = _DeviceInfo;

  factory DeviceInfo.fromJson(Map<String, dynamic> json) => _$DeviceInfoFromJson(json);
}

@freezed
class BreakPeriod with _$BreakPeriod {
  const factory BreakPeriod({
    required String id,
    required DateTime startTime,
    DateTime? endTime,
    required String type, // 'manual' | 'auto' | 'required' | 'geofence_exit'
    Location? location,
    required String triggeredBy, // 'employee' | 'system' | 'admin' | 'schedule'
    int? duration, // minutes, calculated when endTime is set
  }) = _BreakPeriod;

  factory BreakPeriod.fromJson(Map<String, dynamic> json) => _$BreakPeriodFromJson(json);
}

@freezed
class AutoBreakSettings with _$AutoBreakSettings {
  const factory AutoBreakSettings({
    required bool enabled,
    required int requiredBreakDuration, // minutes
    required bool autoStartBreak,
    required bool autoEndBreak,
    required bool geofenceBasedBreaks,
    required bool scheduleBasedBreaks,
    required int minimumWorkBeforeBreak, // minutes
  }) = _AutoBreakSettings;

  factory AutoBreakSettings.fromJson(Map<String, dynamic> json) => _$AutoBreakSettingsFromJson(json);
}

@freezed
class OvertimeSettings with _$OvertimeSettings {
  const factory OvertimeSettings({
    required bool enabled,
    required int thresholdMinutes,
    required bool autoClockOutAtEnd,
    required bool allowOvertime,
    required bool notifyAdminAtOvertime,
    required bool notifyEmployeeAtOvertime,
    required double overtimeRate,
  }) = _OvertimeSettings;

  factory OvertimeSettings.fromJson(Map<String, dynamic> json) => _$OvertimeSettingsFromJson(json);
}

@freezed
class OvertimePeriod with _$OvertimePeriod {
  const factory OvertimePeriod({
    required String id,
    required DateTime startTime,
    DateTime? endTime,
    required int duration, // minutes
    required bool approved,
    String? approvedBy,
    DateTime? approvedAt,
    String? reason,
  }) = _OvertimePeriod;

  factory OvertimePeriod.fromJson(Map<String, dynamic> json) => _$OvertimePeriodFromJson(json);
}

@freezed
class EmployeeNotificationSettings with _$EmployeeNotificationSettings {
  const factory EmployeeNotificationSettings({
    required bool autoTrackingEnabled,
    required bool notifyOnSessionStart,
    required bool notifyOnAutoClockIn,
    required bool notifyOnAutoClockOut,
    required bool notifyOnBreakDetection,
    required bool notifyOnScheduleChange,
    required bool notifyOnOvertimeStart,
    required bool consentGiven,
    required DateTime consentDate,
    required String consentVersion,
  }) = _EmployeeNotificationSettings;

  factory EmployeeNotificationSettings.fromJson(Map<String, dynamic> json) => _$EmployeeNotificationSettingsFromJson(json);
}

@freezed
class CompanyPolicySettings with _$CompanyPolicySettings {
  const factory CompanyPolicySettings({
    required double geofenceAccuracy,
    required int minimumTimeAtSite,
    required bool allowClockInEarly,
    required bool allowClockOutEarly,
    required int clockInBuffer,
    required int clockOutBuffer,
    required double overtimeThreshold,
    required int requiredBreakDuration,
    required int requiredBreakInterval,
    required int geofenceExitGracePeriod,
    String? timeZone,
  }) = _CompanyPolicySettings;

  factory CompanyPolicySettings.fromJson(Map<String, dynamic> json) => _$CompanyPolicySettingsFromJson(json);
}

@freezed
class ScheduleChange with _$ScheduleChange {
  const factory ScheduleChange({
    required String id,
    required DateTime timestamp,
    required String changeType, // 'time_modified' | 'location_changed' | 'cancelled' | 'rescheduled'
    required String changeReason,
    required String changedBy,
    Map<String, dynamic>? oldValues,
    Map<String, dynamic>? newValues,
  }) = _ScheduleChange;

  factory ScheduleChange.fromJson(Map<String, dynamic> json) => _$ScheduleChangeFromJson(json);
}

@freezed
class EventMetadata with _$EventMetadata {
  const factory EventMetadata({
    String? scheduleId,
    DeviceInfo? deviceInfo,
    double? distanceFromJobSite,
    double? accuracy,
    String? source,
    Map<String, dynamic>? additionalData,
  }) = _EventMetadata;

  factory EventMetadata.fromJson(Map<String, dynamic> json) => _$EventMetadataFromJson(json);
}

@freezed
class ScheduleSessionEvent with _$ScheduleSessionEvent {
  const factory ScheduleSessionEvent({
    required String id,
    required DateTime timestamp,
    required String eventType,
    Location? location,
    required String triggeredBy,
    required String details,
    EventMetadata? metadata,
  }) = _ScheduleSessionEvent;

  factory ScheduleSessionEvent.fromJson(Map<String, dynamic> json) => _$ScheduleSessionEventFromJson(json);
}

@freezed
class AdminOverride with _$AdminOverride {
  const factory AdminOverride({
    required String id,
    required DateTime timestamp,
    required String adminId,
    required String adminName,
    required String action,
    required String reason,
    Map<String, dynamic>? originalState,
    Map<String, dynamic>? newState,
  }) = _AdminOverride;

  factory AdminOverride.fromJson(Map<String, dynamic> json) => _$AdminOverrideFromJson(json);
}

@freezed
class SessionError with _$SessionError {
  const factory SessionError({
    required String id,
    required DateTime timestamp,
    required String errorType,
    required String errorMessage,
    required String severity, // 'low' | 'medium' | 'high' | 'critical'
    String? stackTrace,
    Map<String, dynamic>? context,
    required bool resolved,
    String? resolvedBy,
    DateTime? resolvedAt,
    String? resolution,
  }) = _SessionError;

  factory SessionError.fromJson(Map<String, dynamic> json) => _$SessionErrorFromJson(json);
}

@freezed
class ScheduleSession with _$ScheduleSession {
  const factory ScheduleSession({
    required String id,
    required String scheduleId,
    required String employeeId,
    required String employeeName,
    required String jobSiteId,
    required String jobSiteName,
    required String companyId,
    
    // Schedule timing
    String? timeZone,
    required DateTime scheduledStartTime,
    required DateTime scheduledEndTime,
    DateTime? localScheduledStartTime,
    DateTime? localScheduledEndTime,
    
    // Session state
    required String status,
    DateTime? monitoringStarted,
    required bool employeePresent,
    DateTime? arrivalTime,
    DateTime? departureTime,
    DateTime? lastLocationUpdate,
    
    // Time tracking
    required bool clockedIn,
    DateTime? clockInTime,
    DateTime? clockOutTime,
    required bool autoClockInTriggered,
    required bool autoClockOutTriggered,
    
    // Break management
    required bool currentlyOnBreak,
    @Default([]) List<BreakPeriod> breakPeriods,
    required AutoBreakSettings autoBreakSettings,
    
    // Overtime
    required OvertimeSettings overtimeDetection,
    @Default([]) List<OvertimePeriod> overtimePeriods,
    required bool isInOvertime,
    
    // Company settings
    required CompanyPolicySettings companySettings,
    
    // Employee notifications
    required EmployeeNotificationSettings employeeNotifications,
    
    // Events and tracking
    @Default([]) List<ScheduleSessionEvent> events,
    @Default([]) List<AdminOverride> adminOverrides,
    
    // Error handling
    @Default([]) List<SessionError> errors,
    DateTime? lastHealthCheck,
    required String healthStatus,
    
    // Metrics
    required int totalScheduledTime, // minutes
    required int totalWorkedTime, // minutes
    required int totalBreakTime, // minutes
    required int totalOvertimeTime, // minutes
    required double punctualityScore,
    required double attendanceRate,
    required double complianceScore,
    
    // Audit
    required DateTime createdAt,
    required DateTime updatedAt,
    required String createdBy,
    required String lastModifiedBy,
  }) = _ScheduleSession;

  factory ScheduleSession.fromJson(Map<String, dynamic> json) => _$ScheduleSessionFromJson(json);

  // Factory method to create from Firestore document
  factory ScheduleSession.fromFirestore(DocumentSnapshot doc) {
    final data = doc.data() as Map<String, dynamic>;
    return ScheduleSession.fromJson({
      'id': doc.id,
      ...data,
    });
  }
}

// Extension methods for convenience
extension ScheduleSessionExtensions on ScheduleSession {
  
  // Status helpers
  bool get isActive => ['monitoring_active', 'clocked_in', 'on_break', 'overtime'].contains(status);
  bool get isCompleted => status == 'completed';
  bool get isLate => arrivalTime != null && arrivalTime!.isAfter(scheduledStartTime);
  bool get needsAttention => 
      healthStatus == 'error' || 
      status == 'no_show' || 
      isInOvertime || 
      errors.any((e) => !e.resolved);
  
  // Duration calculations
  Duration get currentWorkDuration {
    if (clockInTime == null) return Duration.zero;
    final endTime = clockOutTime ?? DateTime.now();
    final workMs = endTime.millisecondsSinceEpoch - clockInTime!.millisecondsSinceEpoch;
    final breakMs = totalBreakTime * 60000; // Convert minutes to milliseconds
    return Duration(milliseconds: workMs - breakMs);
  }
  
  // Status text helpers
  String get statusText {
    switch (status) {
      case 'scheduled': return 'Scheduled';
      case 'monitoring_active': return 'Monitoring';
      case 'clocked_in': return 'Working';
      case 'on_break': return 'On Break';
      case 'clocked_out': return 'Clocked Out';
      case 'completed': return 'Completed';
      case 'no_show': return 'No Show';
      case 'overtime': return 'Overtime';
      case 'error': return 'Error';
      default: return 'Unknown';
    }
  }
  
  String get statusColor {
    switch (status) {
      case 'scheduled': return 'blue';
      case 'monitoring_active': return 'amber';
      case 'clocked_in': return 'green';
      case 'on_break': return 'orange';
      case 'completed': return 'gray';
      case 'no_show': return 'red';
      case 'overtime': return 'purple';
      case 'error': return 'red';
      default: return 'gray';
    }
  }
}

@freezed
class ScheduleSessionQuery with _$ScheduleSessionQuery {
  const factory ScheduleSessionQuery({
    String? employeeId,
    String? jobSiteId,
    String? companyId,
    String? status,
    DateTime? startDate,
    DateTime? endDate,
    int? limit,
    String? orderBy,
    bool? descending,
  }) = _ScheduleSessionQuery;

  factory ScheduleSessionQuery.fromJson(Map<String, dynamic> json) => _$ScheduleSessionQueryFromJson(json);
}

@freezed
class ScheduleSessionSummary with _$ScheduleSessionSummary {
  const factory ScheduleSessionSummary({
    required String sessionId,
    required String employeeName,
    required String jobSiteName,
    required DateTime scheduledStart,
    required DateTime scheduledEnd,
    required String status,
    required int totalMinutes,
    required double complianceScore,
  }) = _ScheduleSessionSummary;

  factory ScheduleSessionSummary.fromJson(Map<String, dynamic> json) => _$ScheduleSessionSummaryFromJson(json);
}

@freezed
class SessionUpdate with _$SessionUpdate {
  const factory SessionUpdate({
    required String sessionId,
    required String field,
    dynamic oldValue,
    dynamic newValue,
    required DateTime timestamp,
    required String updatedBy,
  }) = _SessionUpdate;

  factory SessionUpdate.fromJson(Map<String, dynamic> json) => _$SessionUpdateFromJson(json);
} 

File: packages/shared/types/schedule-session.ts
// Enhanced Schedule Session Types - Comprehensive Auto Time Tracking System
// Includes break automation, overtime detection, employee consent, and error recovery

export interface BreakPeriod {
  id: string;
  startTime: Date;
  endTime: Date | null;
  type: 'manual' | 'auto' | 'required' | 'geofence_exit';
  location?: {
    latitude: number;
    longitude: number;
    accuracy: number;
  };
  triggeredBy: 'employee' | 'system' | 'admin' | 'schedule';
  duration?: number; // minutes, calculated when endTime is set
}

export interface AutoBreakSettings {
  enabled: boolean;
  requiredBreakDuration: number; // minutes, from CompanySettings
  autoStartBreak: boolean; // automatically start break at scheduled time
  autoEndBreak: boolean; // automatically end break after duration
  geofenceBasedBreaks: boolean; // start break when leaving job site
  scheduleBasedBreaks: boolean; // start break at specific scheduled times
  minimumWorkBeforeBreak: number; // minutes of work before break allowed
}

export interface OvertimeSettings {
  enabled: boolean;
  thresholdMinutes: number; // minutes beyond schedule before overtime
  autoClockOutAtEnd: boolean; // force clock-out when schedule ends
  allowOvertime: boolean; // permit working beyond scheduled hours
  notifyAdminAtOvertime: boolean; // send admin alert when overtime starts
  notifyEmployeeAtOvertime: boolean; // warn employee approaching overtime
  overtimeRate: number; // multiplier for overtime pay calculation
}

export interface OvertimePeriod {
  id: string;
  startTime: Date;
  endTime: Date | null;
  reason: 'schedule_overrun' | 'early_arrival' | 'manual_extension';
  approved: boolean;
  approvedBy?: string;
  duration?: number; // minutes
}

export interface EmployeeNotificationSettings {
  autoTrackingEnabled: boolean;
  notifyOnSessionStart: boolean;
  notifyOnAutoClockIn: boolean;
  notifyOnAutoClockOut: boolean;
  notifyOnBreakDetection: boolean;
  notifyOnScheduleChange: boolean;
  notifyOnOvertimeStart: boolean;
  consentGiven: boolean;
  consentDate: Date;
  consentVersion: string; // for tracking policy updates
}

export interface CompanyPolicySettings {
  geofenceAccuracy: number; // meters
  minimumTimeAtSite: number; // minutes before auto clock-in
  allowClockInEarly: boolean; // clock-in before scheduled time
  allowClockOutEarly: boolean; // clock-out before scheduled end
  clockInBuffer: number; // minutes before schedule start to allow clock-in
  clockOutBuffer: number; // minutes after schedule end to allow clock-out
  overtimeThreshold: number; // hours before overtime calculation
  requiredBreakDuration: number; // minutes
  requiredBreakInterval: number; // hours of work before break required
  geofenceExitGracePeriod: number; // minutes before triggering break/clock-out
}

export interface ScheduleSessionEvent {
  id: string;
  timestamp: Date;
  eventType: 'session_created' | 'monitoring_started' | 'employee_arrived' | 'auto_clock_in' | 
             'manual_clock_in' | 'break_started' | 'break_ended' | 'auto_clock_out' | 
             'manual_clock_out' | 'session_completed' | 'overtime_started' | 'schedule_modified' |
             'admin_override' | 'error_occurred' | 'session_repaired';
  location?: {
    latitude: number;
    longitude: number;
    accuracy: number;
  };
  triggeredBy: 'schedule' | 'employee' | 'system' | 'admin' | 'geofence';
  details: string; // human-readable description
  metadata?: {
    distanceFromJobSite?: number;
    accuracy?: number;
    deviceInfo?: {
      platform: string;
      appVersion: string;
      deviceId: string;
    };
    scheduleChange?: {
      field: string;
      oldValue: any;
      newValue: any;
    };
  };
}

export interface AdminOverride {
  id: string;
  timestamp: Date;
  adminId: string;
  adminName: string;
  action: 'force_clock_in' | 'force_clock_out' | 'start_break' | 'end_break' | 
          'extend_schedule' | 'terminate_session' | 'repair_session' | 'approve_overtime';
  reason: string;
  originalState: any; // snapshot of session before override
  newState: any; // snapshot of session after override
}

export interface SessionError {
  id: string;
  timestamp: Date;
  errorType: 'geofence_failure' | 'location_timeout' | 'schedule_conflict' | 
             'cloud_function_failure' | 'device_offline' | 'permission_denied';
  errorMessage: string;
  severity: 'info' | 'warning' | 'error' | 'critical';
  resolved: boolean;
  resolvedAt?: Date;
  resolvedBy?: string;
  resolution?: string;
}

export interface ScheduleSession {
  // Core session identification
  id: string;
  scheduleId: string;
  employeeId: string; // Keep as employeeId for schedule context
  employeeName: string;
  jobSiteId: string;
  jobSiteName: string;
  companyId: string;

  // Schedule timing with time zone support
  timeZone: string; // Company time zone (e.g., "America/New_York")
  scheduledStartTime: Date; // UTC time
  scheduledEndTime: Date; // UTC time
  localScheduledStartTime: Date; // Local time for display
  localScheduledEndTime: Date; // Local time for display

  // Session state and status
  status: 'scheduled' | 'monitoring_active' | 'clocked_in' | 'on_break' | 
          'clocked_out' | 'completed' | 'no_show' | 'overtime' | 'error' | 'cancelled';

  // Employee presence and arrival tracking
  monitoringStarted: Date;
  employeePresent: boolean;
  arrivalTime: Date | null;
  departureTime: Date | null;
  lastLocationUpdate: Date | null;

  // Time tracking state
  clockedIn: boolean;
  clockInTime: Date | null;
  clockOutTime: Date | null;
  autoClockInTriggered: boolean;
  autoClockOutTriggered: boolean;

  // Break management
  currentlyOnBreak: boolean;
  breakPeriods: BreakPeriod[];
  autoBreakSettings: AutoBreakSettings;

  // Overtime handling
  overtimeDetection: OvertimeSettings;
  overtimePeriods: OvertimePeriod[];
  isInOvertime: boolean;

  // Company policy integration
  companySettings: CompanyPolicySettings;

  // Employee preferences and consent
  employeeNotifications: EmployeeNotificationSettings;

  // Event tracking and audit trail
  events: ScheduleSessionEvent[];
  adminOverrides: AdminOverride[];

  // Error handling and health monitoring
  errors: SessionError[];
  lastHealthCheck: Date;
  healthStatus: 'healthy' | 'warning' | 'error' | 'critical';

  // Calculated metrics and analytics
  totalScheduledTime: number; // minutes
  totalWorkedTime: number; // minutes (excluding breaks)
  totalBreakTime: number; // minutes
  totalOvertimeTime: number; // minutes
  punctualityScore: number; // 0-100, based on arrival vs scheduled time
  attendanceRate: number; // 0-100, based on presence during scheduled time
  complianceScore: number; // 0-100, overall compliance with schedule

  // Audit timestamps
  createdAt: Date;
  updatedAt: Date;
  createdBy: string; // Admin ID who created the schedule
  lastModifiedBy: string; // Last person to modify the session
}

// Supporting types for session queries and operations
export interface ScheduleSessionQuery {
  companyId: string;
  employeeIds?: string[];
  jobSiteIds?: string[];
  startDate?: Date;
  endDate?: Date;
  status?: ScheduleSession['status'][];
  includeErrors?: boolean;
  includeCompleted?: boolean;
  limit?: number;
  offset?: number;
  sortBy?: 'scheduledStartTime' | 'createdAt' | 'updatedAt' | 'punctualityScore';
  sortOrder?: 'asc' | 'desc';
}

export interface ScheduleSessionSummary {
  sessionId: string;
  employeeId: string;
  employeeName: string;
  jobSiteName: string;
  scheduledHours: number;
  workedHours: number;
  breakHours: number;
  overtimeHours: number;
  punctualityScore: number;
  attendanceRate: number;
  complianceScore: number;
  status: ScheduleSession['status'];
  date: Date;
}

// Real-time session update types for live dashboard
export interface SessionUpdate {
  sessionId: string;
  updateType: 'location' | 'status' | 'break' | 'error' | 'overtime';
  timestamp: Date;
  data: any;
}

// Bulk operations for admin management
export interface BulkSessionOperation {
  operationId: string;
  companyId: string;
  type: 'approve_overtime' | 'force_clock_out' | 'repair_errors' | 'extend_sessions';
  sessionIds: string[];
  parameters: any;
  status: 'pending' | 'processing' | 'completed' | 'failed';
  results?: {
    successful: number;
    failed: number;
    errors: string[];
  };
  createdBy: string;
  createdAt: Date;
  completedAt?: Date;
}

// Export as named export only - no default export for types with verbatimModuleSyntax 